fili-core/src/main/java/com/yahoo/bard/webservice/web/util/CacheService.java

```diff
@@ -0,0 +1,173 @@
+    public static final Meter CACHE_REQUESTS = REGISTRY.meter("queries.meter.cache.total");
+
+
+    /**
+     * Read cache.
+     *
+     * @param context The context data from the request processing chain
+     * @param dataCache The cache instance
+     * @param druidQuery The query being processed
+     * @param querySigningService The service to generate query signatures
+     * @param cacheKey Key into which to write a cache entry
+     * @return Response
+     */
+    public String readCache(
+            RequestContext context,
+            TupleDataCache<String, Long, String> dataCache,
+            QuerySigningService<Long> querySigningService,
+            DruidAggregationQuery<?> druidQuery,
+            String cacheKey
+    ) {
+        final TupleDataCache.DataEntry<String, Long, String> cacheEntry = dataCache.get(cacheKey);
+        CACHE_REQUESTS.mark(1);
+
+        if (cacheEntry != null) {
+            if (
+                    querySigningService.getSegmentSetId(druidQuery)
+                            .map(id -> Objects.equals(cacheEntry.getMeta(), id))
+                            .orElse(false)
+            ) {
+                try {
+                    if (context.getNumberOfOutgoing().decrementAndGet() == 0) {
+                        RequestLog.stopTiming(REQUEST_WORKFLOW_TIMER);
+                    }
+
+                    if (context.getNumberOfIncoming().decrementAndGet() == 0) {
+                        RequestLog.startTiming(RESPONSE_WORKFLOW_TIMER);
+                    }
+                    CACHE_HITS.mark(1);
+                    BardQueryInfo.getBardQueryInfo().incrementCountCacheHits();
+                    return cacheEntry.getValue();
+
+                } catch (Exception e) {
+                    LOG.warn("Error processing cached value: ", e);
+                }
+            } else {
+                LOG.debug("Cache entry present but invalid for query with id: {}", RequestLog.getId());
+                CACHE_POTENTIAL_HITS.mark(1);
+                CACHE_MISSES.mark(1);
+            }
+        } else {
+            CACHE_MISSES.mark(1);
+        }
+        return null;
+    }
+
+    /**
+     * Write cache.
+     *
+     * @param response The response handler
+     * @param json Json value to be written to cache as string
+     * @param dataCache The cache instance
+     * @param druidQuery The query being processed
+     * @param querySigningService The service to generate query signatures
+     * @param cacheKey Key into which to write a cache entry
+     * @param writer Writer
+     */
+    public void writeCache(
+            ResponseProcessor response,
+            JsonNode json,
+            TupleDataCache<String, Long, String> dataCache,
+            DruidAggregationQuery<?> druidQuery,
+            QuerySigningService querySigningService,
+            String cacheKey,
+            ObjectWriter writer
+            ) {
```

Comment:
Similarly with the above suggestion:

```suggestion

    /**
     * Write cache.
     *
     * @param response The response handler
     * @param json Json value to be written to cache as string
     * @param druidQuery The query being processed
     */
    public void writeCache(
            ResponseProcessor response,
            JsonNode json,
            DruidAggregationQuery<?> druidQuery
    ) throws JsonProcessingException {
        if (CACHE_PARTIAL_DATA.isOn() || isCacheable(response)) {
            String cacheKey = getKey(druidQuery);

```